### ISSUE TITLE
ClassCastException using OpenJDK 11 (OK in java 8)

### ISSUE THREAD

**Issue by Sciss** [Sciss/SyntaxPane#35]
In GitLab by @pguermo on Dec 4, 2018, 18:50

Migrating from Java 8 to OpenJDK 11, I'm facing a cast exception depending on the JRE used.
The problem can be reproduced using the three attached files.
[pom.xml](/uploads/c2c17d3af3fa0e000d5de94b42286957/pom.xml) [Main.java](/uploads/0e110845d8bfa4c673aefb86dad81801/Main.java) [TextPane.java](/uploads/61d6c8e32b01743783ee46d4e8bf385c/TextPane.java)

```
Exception in thread "main" java.lang.ClassCastException: class javax.swing.text.AbstractDocument$DefaultDocumentEventUndoableWrapper cannot be cast to class javax.swing.text.AbstractDocument$DefaultDocumentEvent (javax.swing.text.AbstractDocument$DefaultDocumentEventUndoableWrapper and javax.swing.text.AbstractDocument$DefaultDocumentEvent are in module java.desktop of loader 'bootstrap')
	at de.sciss.syntaxpane.CompoundUndoManager.undoableEditHappened(CompoundUndoManager.java:62)
	at java.desktop/javax.swing.text.AbstractDocument.fireUndoableEditUpdate(AbstractDocument.java:293)
	at java.desktop/javax.swing.text.AbstractDocument.handleInsertString(AbstractDocument.java:761)
	at java.desktop/javax.swing.text.AbstractDocument.insertString(AbstractDocument.java:716)
	at java.desktop/javax.swing.text.PlainDocument.insertString(PlainDocument.java:131)
	at java.desktop/javax.swing.text.DefaultEditorKit.read(DefaultEditorKit.java:274)
	at de.sciss.syntaxpane.DefaultSyntaxKit.read(DefaultSyntaxKit.java:217)
	at java.desktop/javax.swing.JEditorPane.setText(JEditorPane.java:1428)
	at com.testing.Main.<init>(Main.java:8)
	at com.testing.Main.main(Main.java:13)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
	at com.intellij.rt.execution.application.AppMainV2.main(AppMainV2.java:131)
```

**Comment by Sciss** [comment]
In GitLab by @pguermo on Dec 4, 2018, 19:03

This problem seems to be quite generic. See https://netbeans.org/bugzilla/show_bug.cgi?id=270142

> This bug i due to a change in Java 9 where the edit of the argument (a 
> DefaultDocumentEvent) is first wrapped in a DefaultDocumentEventUndoableWrapper which hides the line number and other important stuff (I can't understand why!!), before 
> 
>    public void undoableEditHappened(UndoableEditEvent e)
> 
> is invoked.
> 
> https://github.com/netroby/jdk9-dev/blob/master/jdk/src/java.desktop/share/classes/javax/swing/text/AbstractDocument.java#L279
> 
> This makes this cast break:
> 		AbstractDocument.DefaultDocumentEvent event =
> 			(AbstractDocument.DefaultDocumentEvent)e.getEdit();
> 
> The problem is that this cast is widely used in all kinds of code - originating from this post from 2008: https://tips4java.wordpress.com/2008/10/27/compound-undo-manager/
> 
> Here is a possible fix:
> https://github.com/nordfalk/jsyntaxpane/commit/5fc75594f8bc4df6e8f7096d4a440490b768fd46#diff-b3f8826824f551219879f1c798618c12R67

**Comment by Sciss** [comment]
In GitLab by @pguermo on Dec 4, 2018, 19:07

changed the description

**Comment by Sciss** [comment]
In GitLab by @pguermo on Dec 5, 2018, 10:42

Here is a patch [git.diff](/uploads/6f600405ad468dbb3805c10592d6fe08/git.diff). 
I've tested it with success with my application.

I hope a new version will be released quite soon to benefit of the update. Thanks ;)

**Comment by Sciss** [comment]
In GitLab by @Sciss on Dec 10, 2018, 12:33

Hi Patrick, sorry I had wrong notification settings and didn't see this before. Thank you for the patch, I will have a look into it.

**Comment by Sciss** [comment]
In GitLab by @Sciss on Jan 8, 2019, 16:57

This was fixed in 85fdde1a - published version 1.2.0

**Comment by Sciss** [comment]
In GitLab by @Sciss on Jan 8, 2019, 16:57

closed
